### ISSUE TITLE
Intermittent "NoSuchKey: The specified key does not exist" errors when using S3 backend

### ISSUE THREAD

**Issue by rtkelly13** [pulumi/pulumi#2714]
I have been following the cloud backend change

https://github.com/pulumi/pulumi/pull/2455

When trying to run
`pulumi login s3://pulumi-state-files` and then `pulumi up`

I get the following error 
`error: failed to load checkpoint: blob (code=NotFound): NoSuchKey: The specified key does not exist.`
I can navigate and interrogate that specific S3 bucket via the AWS CLI and have setup the appropriate AWS_PROFILE environment.

Not quite sure what I am doing incorrectly, I have followed all the documentation and issue steps I have been able to find.

**Comment by CyrusNajmabadi** [comment]
Unfortunately, i don't repro this at all.  Here's what i did:

1. from AWS i made a bucket called cytestbucket2 (with all public access blocked).
2.  then did a: `pulumi login s3://cytestbucket2`, which gave me:

```
cyrusn@DESKTOP-3IRRNND ~/go/src/github.com/pulumi/pulumi-aws/examples/topic[master ≡]$2> pulumi login s3://cytestbucket2
Logged into DESKTOP-3IRRNND as cyrusn (s3://cytestbucket2)
```

3. then did a: `pulumi up` which gave me:

cyrusn@DESKTOP-3IRRNND ~/go/src/github.com/pulumi/pulumi-aws/examples/topic[master ≡]$2> pulumi up -- skip-preview
Updating (cytesttopic):

     Type                                Name                    Status
 +   pulumi:pulumi:Stack                 topic-cytesttopic       created
 +   ├─ aws:sns:TopicEventSubscription   for-each-url            created
 +   │  ├─ aws:iam:Role                  for-each-url            created
 +   │  ├─ aws:iam:RolePolicyAttachment  for-each-url-32be53a2   created
 +   │  ├─ aws:lambda:Function           for-each-url            created
 +   │  ├─ aws:sns:TopicSubscription     for-each-url            created
 +   │  └─ aws:lambda:Permission         for-each-url            created
 +   └─ aws:sns:Topic                    sites-to-process-topic  created

Resources:
    + 8 created

Duration: 24s

Permalink: <redacted>
```

This all seemed to work as desired.  

Since this is not working for you, we'll need more information to track things down.  Can you give me a set of steps you followed that caused the above errors?

Thanks!

**Comment by CyrusNajmabadi** [comment]
Explicitly pinging @rtkelly13 to help ensure the above is seen.  Thanks!

**Comment by lukehoban** [comment]
> I can navigate and interrogate that specific S3 bucket via the AWS CLI and have setup the appropriate AWS_PROFILE environment.

It sounds like this part may be key to reproing this - specifically the use of an `AWS_PROFILE` environment variable to select the profile to use.

**Comment by CyrusNajmabadi** [comment]
Will try to validate that scenario as well.

**Comment by CyrusNajmabadi** [comment]
Unfortunately, was still not able to repro a problem here.  Note: when my profile information was unspecified/incorrect, i got messages of the form:

```
error: failed to load checkpoint: blob (code=Unknown): NoCredentialProviders: no valid providers in chain. Deprecated.
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

However, if i setup my credentials to have a non-default profile (called `user`), and i did `export AWS_PROFILE=user1`, then things worked totally fine.  

I'm also going to try the general approach of fully specifying things through env-vars to see if that works as well.

**Comment by CyrusNajmabadi** [comment]
Note: we def want to get to the bottom of things to help all of you out.   So any extra info/steps that can be provided would be great.

**Comment by CyrusNajmabadi** [comment]
Additional info.  If the bucket itself doesn't exist, i get this different error:

```
cyrusn@DESKTOP-3IRRNND ~/go/src/github.com/pulumi/pulumi-aws/examples/topic[master ≡]$2> yarn run build && pulumi up
yarn run v1.15.2
$ tsc
Done in 7.05s.
error: could not query backend for stacks: error listing stacks: could not list bucket: blob (code=Unknown): NoSuchBucket: The specified bucket does not exist
        status code: 404, request id: 3C2CABEED8636B03, host id:
```

So def scratching head as to what's causing the `NoSuchKey: The specified key does not exist.` error.  Will continue to look into things.

**Comment by CyrusNajmabadi** [comment]
Hey @rtkelly13 I think i may have uncovered what you're hitting.  Specifically, i was able to get that error message when i did the following:

1. Be logged into the normal pulumi service backend (i.e. a normal `pulumi login` without an s3 bucket backend.
2. Created and updated my stack.
3. Logged out of the normal backend, then logged into the `s3` backend.
3. Updated my stack.

In this case, pulumi thinks the stack (and it's requisite info) exists somewhere, and it tries to load it, but it can't find it because it's now looking into the `s3` bucket for the information.

First, can you confirm if this sounds like what you were encountering?  If so, we can start brainstorming on what we want to happen here.  At the very least, the error message should bcome clear as to what the problem is.  If we want to take further work beyond that is TBD.

**Comment by rtkelly13** [comment]
Thank you for the work on this!

I have been using the local file backend but could be a similar cause to your reproduction, I wouldn't have expected a pre-existing stack on a different login to side effect in this way? It's good to know that the bucket and authentication seem to be working correctly for me. 

I am using Windows also if that may help in my specific reproduction, I guess I should be importing my stack into S3 before trying to sync up with my existing stack? I think it was just that error that threw me off course, as you said better error messages should avoid further issues being created.

I'm also not a huge fan of setting the AWS_PROFILE to say what credentials to use, it seems to be the only way AWS provides unfortunately.

**Comment by CyrusNajmabadi** [comment]
> I am using Windows also if that may help in my specific reproduction, I guess I should be importing my stack into S3 before trying to sync up with my existing stack? 

That sounds reasonable.  It's possible (though not verified) that logging into the right backend, exporting your existing stack, logging out, logging into the new backend, and importing the stack might work.  If it does, let us know.  

> I think it was just that error that threw me off course, as you said better error messages should avoid further issues being created.

Agreed.  Will see what can be done here.

> I'm also not a huge fan of setting the AWS_PROFILE to say what credentials to use, it seems to be the only way AWS provides unfortunately.

How many profiles do you use?  I believe you can set a `[default]` in your credentials file which AWS will respect.

**Comment by CyrusNajmabadi** [comment]
> I guess I should be importing my stack into S3 before trying to sync up with my existing stack

Note: I think primarily we've been considering cloud-backends as a way for people to store their new stacks.  Migration between backends hasn't really been something we've thought about or considered as something that needs to be supported.  It may work, or it may not.  I think we have the primitives to help support the workflows you want.  But once you start trying to juggle all this, it may be necessary on your end to make a little sausage to get through some bits :)

**Comment by rtkelly13** [comment]
Tried giving importing the stack into the S3 backend but got the same error as before
I had already exported the state file into source control to avoid losing it if something happened to my environment.

`error: failed to load checkpoint: blob (code=NotFound): NoSuchKey: The specified key does not exist.`

Using the following command.
`pulumi stack import --file .\{STATEFILENAME}.json`

Going to give creating a brand new stack a go with S3 from the start, will give me an idea of the file structure that is uploaded and will give me an idea of how to fudge it, hopefully I can do it pretty easily and will put reproduction steps into here for anyone else who may need to do this.

Going to keep playing with this as it isn't a blocker for me personally, I know of a few issues around this area that I haven't fully reproduced enough to create issues for them.

> How many profiles do you use? I believe you can set a [default] in your credentials file which AWS will respect.

default would be respected, I tend to have a few set-up especially if I am switching between different AWS accounts, via local development or CI agents it a good way to swap out credentials for each stage of the deployment process

>>I think it was just that error that threw me off course, as you said better error messages should avoid further issues being created.

> Agreed. Will see what can be done here.
Knowing what key it is complaining about would massively help me massively when manually trying to migrate my stack over to S3 and I'm sure other people in my situation with any of the clouds

**Comment by CyrusNajmabadi** [comment]
> Going to give creating a brand new stack a go with S3 from the start, will give me an idea of the file structure that is uploaded and will give me an idea of how to fudge it, hopefully I can do it pretty easily and will put reproduction steps into here for anyone else who may need to do this.

I **think** (but am not totally certain) that part of this is due to another piece of data we store *locally* to keep track of which stack you're currently using.  That information is normally under `~/.pulumi/workspaces/name_hash...`.  You might need to delete that so that we don't have this data sticking around that is causing us to think yo have a currently selected stack.

Note: it's def possible we need to fixup several thigns here.  i.e. that logging out might need to do this, or that the stored stack information knows which backend you're using and it will not apply if it doesn't correspond to your logged in backend.  I'll have to discuss this with colleagues to decide what's likely the best approach here.

**Comment by CyrusNajmabadi** [comment]
Moving to 0.24 as this is more about a better error message for this scenario.

**Comment by rtkelly13** [comment]
The way I got this working was to copy the appropriate stack files from ~/.pulumi/stacks 
Which for me were `STACKNAME.json` and `STACKNAME.json.bak` into my S3 bucket at the same place. I ignored  `STACKNAME.json.attrs` and `STACKNAME.json.bak.attrs` and it works without them!

After that, I ran `pulumi stack select STACKNAME` which worked and I updated the stack and that worked as expected. If I had just used S3 from the beginning it would have been a very good experience!

I think there are still a few rough edges probably from the change to support this cloud backend feature
`pulumi stack ls` doesn't work to print out the available stacks.
 
When I interrogate the stack it still says it is managed by my local desktop, which is obviously because I copied the files, if an update changed that information to the S3 path given that might be better?
Could help detect the change in login etc?

It seems like most of the things required to interact with a specific stack just work! As you said its more about user experience and failed commands guiding users in the right direction the first time.

**Comment by CyrusNajmabadi** [comment]
Glad to hear you were able to work through things!

> It seems like most of the things required to interact with a specific stack just work! As you said its more about user experience and failed commands guiding users in the right direction the first time.

Indeed.  Our primary goal was to just enable this high value scenario for users.  Secondary is smoothing out these rough edges, esp. for cases like migrating stacks.  Def something we'd like to do, but has to be prioritized against all the other work for the forseeable future :)

**Comment by rtkelly13** [comment]
@CyrusNajmabadi thank you for your help, really like pulumi and the approach taken!
Should this issue be closed and a more appropriate one be opened in relation to the outstanding issues?

**Comment by CyrusNajmabadi** [comment]
I'm ok keeping this issue open.  It's a good documentation of the problem.  If we opened a new issue, we'd just have to link back to this for whoever works on it to understand what's going on and get teh necessary context.

Thanks!

**Comment by keen99** [comment]
```Be logged into the normal pulumi service backend (i.e. a normal pulumi login without an s3 bucket backend.
Created and updated my stack.
Logged out of the normal backend, then logged into the s3 backend.
Updated my stack.
```

now that AWS_PROFILE support is fixed in 0.17.15, and the project I'm working on is ready to try to manage multiple accounts, I've started running into workflow issues around this as well.

it seems like there might be a strong case going forward to be able to track `login` destination in the stack.yaml.     then during any stack level interaction a login could happen automatically to the apropos destination.

a good example of this in an s3-state-only world:   i have 3 stacks.  `dev` `stage` `prod`.  each of these lives on their own account.   that means they require their own s3 bucket (with a globally unique name), and their own `login` step each time you decide to operate on the stack.

so instead of a promotion workflow like this:
```
pulumi up -s dev
pulumi up -s stage
pulumi up -s prod
```

I end up like this:
```
export AWS_PROFILE=dev
pulumi login s3://dev-bucket/
pulumi up -s dev
pulumi logout

export AWS_PROFILE=stage
pulumi login s3://stage-bucket/
pulumi up -s stage
pulumi logout

export AWS_PROFILE=prod
pulumi login s3://prod-bucket/
pulumi up -s prod
pulumi logout
```

now multiply that by a simple client with 3 or 4 products, and 2 or 3 clients in parallel, and you've got a strong requirement to start building a wrapper around pulumi just to handle pulumi....   I already have a strong drive to do that because of how `current stack` is stored that prevents operating on multiple stacks concurrently (or adjacent in different shells).

**Comment by keen99** [comment]
I should note that a single bucket can handle all of the stage cross-accounts.

```
export AWS_PROFILE=... 
export AWS_REGION=.. 
pulumi login s3://bucket
```

other pulumi up etc operations will follow this - but if you set stack config `aws:profile: newprofile` the stack itself will follow the stack's profile, not the env var (at least in my light testing).   state will still store based on the env var and login.

**Comment by lukehoban** [comment]
cc @ellismg as the issues here really highlight that it might make sense to tie backend more closely to specific stacks - or at least remember (in `~/.aws/workspaces`?) which backend a stack was created in so that at the very least a helpful error message can be reported?

**Comment by keen99** [comment]
> or at least remember (in ~/.aws/workspaces?)

then it's a piece of config that doesnt follow the code since it can't be commited.    (though I'm totally fine if that's standard but you can elect to commit it).

for sure this is a change from the pulumi SaaS backend only, but it seems critical now that there's more than a single choice to treat this as config for the code.

**Comment by keen99** [comment]
> you've got a strong requirement to start building a wrapper around pulumi just to handle pulumi
...and I wrote it.   there's no native MFA handling for aws auth that requires mfa....

**Comment by ellismg** [comment]
I understand the pain here, and I agree that we can probably figure out something better.  The thing that I am struggling with is how `pulumi stack ls` operates in the is new world.  There are two modes today:

`pulumi stack ls`

This this lists all the stacks in the same project from the current backend.  It must be run in a folder with a `Pulumi.yaml` either in it or in a parent folder.  You see stacks for that project even if you don't have the config file for them on disk.

`pulumi stack ls --all` 

This lists **all** stacks in the current back-end.  So you see every stack you have access to, even if you don't have the source code for their corresponding projects around.

If we move to a world where we use information in the specific `Pulumi.<stack-name>.yaml` file to help determine the backend, does that mean we lose the ability for `pulumi stack ls --all` to work?  Does `pulumi stack ls` now only show the set of stacks that have configuration on disk?

**Comment by keen99** [comment]
it seems like this would be functionality you'd lose, or have limitations on, when using separate backends.

that said - it might be a nice touch to be able to have the state backend auth separately from a stack, so you could designate an AWS account+bucket that's independent of where the stack ends up.   that would leave those features intact, and retain similar behavior to the pulumi backend.

I do think that having a per-stack backend should be an optional workflow.  


the way I've structured my current wrapper, it enforces a `-s/--stack` be passed (with a few exceptions), then it consumes `aws:profile` and `pulumi:backend` from the stack.yaml, then externally auths to aws (assume role, prompting for MFA as required - in the future, this would do SAML+MFA), then `login pulumi:backend`.  then it'll pass everything to the real `pulumi`.    there will definitely be aspects of the overall system that probably wont sane here.

for the workflow I see working for the client this is for, this makes sense so far.   (imperfect, but that's life)

**Comment by bdchauvette** [comment]
I've started running into this problem as well. Unfortunately, it's intermittent and I haven't been able to figure out what causes it. It does seem to have started happening once my state file got larger (>1mb?).

Sometimes `pulumi up` will work perfectly fine, and other times it will blow up with either pre-step or post-step errors like:

```
 error: post-step event returned an error: failed to save snapshot: failed to load checkpoint: blob (code=NotFound): NoSuchKey: The specified key does not exist.
    status code: 404, request id: [...], host id: [...]
```

```
error: pre-step event returned an error: failed to save snapshot: failed to load checkpoint: blob (code=NotFound): NoSuchKey: The specified key does not exist.
    status code: 404, request id: [...], host id: [...]
```

The stack is trying to spin up an EKS cluster with a couple node groups, and then create some Kubernetes resources in the cluster. The AWS infrastructure steps seem to work OK, and it seems like the errors usually happen when creating the Kubernetes resources?

---

```
$ pulumi version
v0.17.25
```

```json
"dependencies": {
  "@pulumi/aws": "0.18.23",
  "@pulumi/awsx": "0.18.7",
  "@pulumi/eks": "0.18.9",
  "@pulumi/kubernetes": "0.25.2",
  "@pulumi/pulumi": "0.17.25"
},
```

**Comment by lukehoban** [comment]
@bdchauvette which backend are you using?  S3?

**Comment by rico1342** [comment]
@

**Comment by Mohsenam99** [comment]
bitcoin:3LosJGMEsPnrBMnpWPbSHrTPGisueLiAFD?amount=0.00090529&label=Your%20subscription%20to%20Bitcoin%20M
